backend: pkg: portforward: Fix cache implementation and refactor shorten PortForward

fromsaurav opened the pull request:
Refactor(backend/portforward): Shorten startPortForward function

This PR addresses issue #3279 by refactoring the internal `startPortForward` function in `backend/pkg/portforward/handler.go`.

**Changes Made:**
- The `startPortForward` function was broken down into smaller, more manageable helper functions:
    - `getKubeClientAndConfig`
    - `initPortForwarder`
    - `handlePortForwardReadiness`
    - `runAndMonitorPortForward`
    - `monitorPodAndManagePortForward`
- This refactoring resolves the `funlen` linter warning previously associated with this function.
- The primary goal was to improve code readability, maintainability, and reduce complexity without altering existing functionality.
- The `//nolint:funlen` directive has been removed from the refactored `startPortForward` function.

**Testing:**
- All backend linters (`make backend-lint`) pass.
- All backend unit tests (excluding the portforward integration test's specific content assertions) pass.
- The integration test `TestStartPortForward` was run with `HEADLAMP_RUN_INTEGRATION_TESTS=true` against a Minikube cluster using a simple Nginx pod in the `headlamp` namespace:
    - The test successfully established the port-forward connection to the Nginx pod (as confirmed by logs from the handler and successful HTTP connection to the forwarded port).
    - The test currently fails on application-level assertions (checking for `/config` endpoint content and "incluster" string) because the Nginx pod does not serve the Headlamp application. This is an expected outcome with the generic test pod.
- The core port-forwarding setup, lifecycle management, and teardown logic refactored in this PR appear to be functioning correctly.
- Relying on CI for full validation of `TestStartPortForward` against a complete Headlamp deployment where application-level assertions should pass.

Fixes #3279

illume commented:
We follow a linux kernel style of commit messages. With the context at the front followed by an action word and the rest.

Would you mind changing them to something like this? (but probably you can write them better). 

`backend: portforward: Function: Shorten function`


See git log for some more examples.

illume commented:
@fromsaurav there is a conflict now in this PR. Would you mind rebasing against the latest main and fixing the conflict?

fromsaurav commented:
> Thank you. I left some notes.
> 
> It looks like a bunch of separate changes in here. Are you able to break this up into different parts with a different commit for each change? This will make it easier to follow, and make debugging later easier if there is a problem. The changes should be atomic commits where each commit passes all checks and makes sense on its own.

@illume , made some changes, 
-added docs for functions
-removed comments
-made atomice comments
is that fine?

illume commented on backend/pkg/portforward/handler_test.go:
```diff
@@ -67,49 +67,69 @@ func TestStartPortForward(t *testing.T) {
-	err = kubeConfigStore.AddContext(&kc)
+	const minikubeName = "minikube"
+
+	// WSL: declarations should never be cuddled
```
Thanks

illume commented on backend/pkg/portforward/handler.go:
```diff
@@ -192,105 +194,256 @@ func StartPortForward(kubeConfigStore kubeconfig.ContextStore, cache cache.Cache
-func startPortForward(kContext *kubeconfig.Context, cache cache.Cache[interface{}],
-	p portForwardRequest, token string,
-) error {
+func getKubeClientAndConfig(kContext *kubeconfig.Context, token string) (*kubernetes.Clientset, *rest.Config, error) {
```
thanks

illume reviewed (approved):
Very nice, thanks!
